[docs] Add `expo-server` docs to API routes page (#40521)
expo/expo: docs/pages/router/reference/api-routes.mdx

## Requests

> <a id="requests" />

Author: Oh, I was about to rename this heading and then didn't 🙃 I'll remove it

Making requests with an undefined method will automatically return `405: Method not allowed`. If an error is thrown during the request, it will automatically return `500: Internal server error`.

## Runtime API

> **important** The server runtime API and `expo-server` are available in SDK 54 and later, and require a deployed server for production use.

You can use the `expo-server` package to use several utilities and code patterns that work in any server-side Expo code. This includes utilities to get request metadata, for scheduling tasks, and for error handling.

<Terminal cmd={['$ npx expo install expo-server']} />

> Using `expo-server` is not limited to API routes and it can be used in any other server code as well, for example in [server middleware](/router/reference/middleware/).

Reviewer: ```suggestion
Using `expo-server` is not limited to API routes and it can be used in any other server code as well, for example, in [server middleware](/router/reference/middleware/).
```

export async function GET(request: Request, { post }: Record<string, string>) {
  if (!post) {
    throw new StatusError(404, 'No post found');
  }
  // ...
}
```

When composing your own server utilities and helpers, the `StatusError` is a more convenient way to handle exceptions, since throwing them interrupts any API functions and instead returns an error early.

`StatusError`s accept a status code and an error message, which can also optionally be passed as a JSON, or `Error` object, and will always return a `Response` with a JSON body with an `error` key set to their error message.

This can be restrictive, and isn't suitable for all cases. Sometimes it might be beneficial to instead `throw` a `Response` object, which interrupts your logic as well, but replaces the resolved `Response` from your API route directly, without a `StatusError` wrapper.
For example, this can be used to create redirect responses.

```ts app/blog/[post]+api.ts
import { StatusError } from 'expo-server';

export async function GET(request: Request, { post }: Record<string, string>) {
  if (!post) {
    throw Response.redirect('https://expo.dev', 302);
  }
  // ...
}
```

### Request metadata

> Requests typically carry most metadata you'll need in their headers. However, `expo-server` provides some helpers functions to retrieve common values more easily.

Reviewer 2: ```suggestion
Requests typically carry most metadata you'll need in their headers. However, `expo-server` provides some helper functions to retrieve common values more easily.
```

Helper functions from `expo-server` return values that are scoped to the current `Request`. You can only call these functions in server-side code and only during ongoing requests.

> A common value that you may need to access is the request's origin URL. The origin URL - typically transmitted on a request's `Origin` header, represents the URL that a user used to access your API route. This may differ from any internal deployment URL that your server sees when the request is being proxied. You can use `expo-server`'s [`origin()`](/versions/latest/sdk/server/#origin) helper to access this value.

Reviewer: ```suggestion
A common value that you may need to access is the request's origin URL. The origin URL, typically transmitted on a request's `Origin` header, represents the URL that a user used to access your API route. This may differ from any internal deployment URL that your server sees when the request is being proxied. You can use `expo-server`'s [`origin()`](/versions/latest/sdk/server/#origin) helper method to access this value.
```

You can use `expo-server`'s [`deferTask()`](/versions/latest/sdk/server/#defertaskfn) helper to schedule tasks to run after a `Response` has been resolved by your API route.

```ts app/tasks+api.ts
import { deferTask } from 'expo-server';

export async function GET(request: Request) {
  // This will run after this entire function resolves:
  deferTask(async () => {
    await pingAnalytics(...);
  });

  const data = await fetchExampleData(...);
  return Response.json({ data });
}
```

### Response headers

When structuring and splitting server logic into separate helper functions and files, it may be necessary to modify `Response` headers before a `Response` has been created.

For example, you may need to add metadata in [server middleware](/router/reference/middleware/) to a `Response` before your API route code is running.

```ts app/+middleware.ts
import { setResponseHeaders } from 'expo-server';

export default function middleware(request: Request) {
  // Rate limiters typically add a `Retry-After` header
  setResponseHeaders({ 'Retry-After': '3600' });
}
```

> In the above example, we're adding a `Retry-After` header to a future `Response` that an API route may be creating. This can also be extended for authentication and cookies.

Reviewer: ```suggestion
In the above example, a `Retry-After` header is added to a future `Response` that an API route may be creating. This can also be extended for authentication and cookies.
```

- `expo-server` does **not** inflate environment variables from **.env** files. They are expected to load either by the hosting provider or the user.
- Metro is not included in the server.

> The `expo-server` package contains adapters for various providers and runtimes. Before proceeding with any of the below sections, install the `expo-server` package first.

Reviewer: ```suggestion
The `expo-server` library contains adapters for various providers and runtimes. Before proceeding with any of the below sections, install the `expo-server` library.
```
